4325: Config import (yaml)

cableman opened the pull request:
#### Link to issue

https://platform.dandigbib.org/issues/4925

#### Description

Import configuration via YAML file upload (with/without flattening the arrays, mostly for readability for PL's). The modules comes with an YAML example file and readme about the modules usage.

#### Screenshot of the result

![Screenshot 2020-12-08 at 12 30 04](https://user-images.githubusercontent.com/111397/101478533-202ebc00-3951-11eb-80b2-0cf8ce73daa4.png)

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+function _ding_config_import_set_values(array $values, $prefix = '') {
+  foreach ($values as $key => $value) {
+    switch ($key) {
+      case 'modules':
+        // Disable modules.
+        if (!empty($value['disable'])) {
+          module_disable($value['disable']);
+        }
```
I do not think it is a good idea to have module enabling/disabling as a part of the recursive function. It increases the complexity of the function and can lead to problems if a key within an array intended for variables is called `modules`.

I suggest using separate root elements `variables`/`settings` and `modules` and using separate functions for handling each of them. This also allows for further expansion down the line.

cableman replied:
Updated to have root elements `settings` and `modules`

kasperg replied:
:+1: I think you should update the settings examples in the documentation accordingly.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+        break;
+
+      default:
+        if (is_array($value) && _ding_config_import_is_associative($value)) {
+          if (!isset($value['array'])) {
+            _ding_config_import_set_values($value, _ding_config_import_prefix($prefix, $key));
+            break;
+          }
+        }
+
+        if (isset($value['array']) && $value['array'] == 1) {
+          unset($value['array']);
+        }
```
```suggestion
        if (is_array($value) && _ding_config_import_is_associative($value)) {
          if (!isset($value['array']) || $value['array'] != 1  ) {
            _ding_config_import_set_values($value, _ding_config_import_prefix($prefix, $key));
            break;
          } else {
            unset($value['array']);          
          }          
        }
```

I found the multiple conditionals hard to grok. I suggest changing to this as the outer if-statement tests for associative arrays. That has to be true before checking for whatever value the `array` entry might be.

cableman replied:
Suggestion added

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+            // Can not throw an exception here as that would stop the other
+            // changes from being applied. So use set message directly here to
+            // notify the use that module enable failed.
+            drupal_set_message(t('Not all module was enabled check the name and the drupal logs for errors.'), 'error');
```
```suggestion
            drupal_set_message(t('Not all modules were enabled check the name and the drupal logs for errors.'), 'error');
```

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+/**
+ * Recursive walk the array using the key as prefix.
+ *
+ * If this an multilevel array the keys are joined with "_". If the flatten
+ * parameter is FALSE.
```
The function does not have a `flatten` parameter anymore. Is this still relevant?

cableman replied:
No, it's an left over form first internal iterration. So comment removed

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+    'wrapper' => [
+      '#type' => 'fieldset',
+      '#title' => t('Configuration upload'),
+      '#description' => t('This module comes with an <a href="/@url">example file</a> that can be used as an template for the upload. Note that enabled/disable modules may timeout, if it takes longer than the web-servers configuration allows.', ['@url' => $url]),
```
It is nice with the notice. We could consider handling this using Batch API.

cableman replied:
I think that should be an improvement if this turns out to be a problem.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+  $url = drupal_get_path('module', 'ding_config_import') . '/example.config.yml';
+
+  $form = [
+    '#attributes' => [
+      'enctype' => 'multipart/form-data',
+    ],
```
The documentation says this is not needed as [Drupal Form API will add it automatically](https://api.drupal.org/api/drupal/developer%21topics%21forms_api_reference.html/7.x#file).

cableman replied:
Check :-)

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+    }
+  }
+  else {
+    $error = true;
+    $msg = t('Fail to load YAML parsing library. Please contact the system administrator.');
+  }
+
+  if ($error) {
+    drupal_set_message($msg, 'error');
```
If you change `$error` to `$type` and from a boolean to a string you can leave out the final conditional and just call `drupal_set_message()` with the parameters.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+  $file = $_FILES['files']['tmp_name']['upload'];
+
+  if (libraries_load('yaml')) {
+    $data = file_get_contents($file);
+    $values = Yaml::parse($data);
```
```suggestion
    $values = Yaml::parseFile($file);
```

This is shorter and [you also get better error handling from the Yaml component regarding file access](https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Yaml/Parser.php).

kasperg replied:
You :+1: this but I do not think your changes will work. You must pass the filename to `Yaml::parseFile()` - not the file contents.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+  if (libraries_load('yaml')) {
+    $data = file_get_contents($file);
+    $values = Yaml::parse($data);
+    if (is_array($values)) {
+      _ding_config_import_set_values($values);
+    }
+    else {
+      $error = true;
+      $msg = t('Fail to parse the YAML file.');
+    }
```
As I read it the YAML parser throws exceptions if problems occur. If we want to introduce error handling then we should rely on that.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+function ding_config_import_menu() {
+  $item = [];
+
+  $item['admin/config/ding/config'] = [
```
```suggestion
  $item['admin/config/ding/config_import'] = [
```

This makes the path follow the module name. `config` is very generic.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+    'description' => 'Upload and import configuration from YAML.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_config_import_admin_form'),
+    'access arguments' => array('administer site configuration'),
```
I wonder if `administer site configuration` is sufficiently strict for this?

The module effectively allows you to bypass access control for other configuration pages which may have their own access control settings which are more strict that `administer site configuration`.

We could also consider [marking the module as `hidden`](https://www.drupal.org/docs/7/creating-custom-modules/writing-module-info-files-drupal-7x#hidden) to avoid someone enabling it in a production environment.

cableman replied:
I see your point and will add hidden, but it may give issues on the non-platform.sh test sites.

kasperg replied:
I see your point.

With the approach you will have to use Drush to enable the module. Testers will usually have to talk to a developer in these cases.

Another option would be to restrict access to user 1. Perhaps that is more feasible?

kasperg commented on project.make:
```diff
@@ -674,3 +674,9 @@ libraries[fabric][download][type] = "get"
 libraries[fabric][directory_name] = "fabric"
 libraries[fabric][destination] = "libraries"
+
+; Library used by Configuration import (older version to be able to run on PHP 5.6).
```
:+1: for explaining this.

kasperg commented on modules/ding_config_import/README.md:
```diff
@@ -0,0 +1,74 @@
+The configuration is described by using YAML files with arrays that are flattened to variable keys
+if they are associative arrays. If "normal" arrays they are written as arrays to variables.
+
+If an associative array contains the stop key `array: 1` the array it is placed in is not flattened
+thereby allow for complex variable configuration.
```
:+1: for this. Perhaps consider prefixing the key with a special character to make it stand out more? E.g. `_array`. You might also consider using [a proper boolean value](https://yaml.org/type/bool.html) instead of a numeric value.

kasperg commented on modules/ding_config_import/README.md:
````diff
@@ -0,0 +1,74 @@
+
+Upload the file at `/admin/config/ding/config` and clear the cache.
+
+## How it is parsed
+
+This yaml snippet will be transformed to the following variables as store in the database.
+```yaml
+---
+opensearch:
+  url: "https://opensearch.addi.dk/b3.5_5.2/"
````
These examples are excellent. I wish all our READMEs could be like this :smile:.

kasperg commented on modules/ding_config_import/ding_config_import.info:
```diff
@@ -0,0 +1,4 @@
+name = Ding Config Import
+description = Import configuration from YAML file
+core = 7.x
+package = Ding!
```
Please remember to include the dependency on `ding_libs`.

cableman replied:
Added

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+    drupal_set_message($msg, 'status');
+  }
+
+  // Ensure tmp file is removed. It contains secrets.
```
```suggestion
  // Ensure tmp file is removed. It may contain secrets.
```

cableman replied:
Fixed

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,169 @@
+ */
+function ding_config_import_admin_form_submit($form, &$form_state) {
+  $error = false;
+  $msg = t('The configuration options have been saved. You may need to clear cache to rebuild things if you have enabled/disabled modules etc.');
```
This is somewhat vague 😅 . We could consider just always clearing the cache after an import. However this might increase the risk of timeouts - especially without Batch API.

cableman replied:
yes, therefor an improvement for later.

kasperg reviewed (changes requested):
Lots of good stuff here and lots of comments :smile:.

cableman commented on modules/ding_libs/ding_libs.module:
```diff
@@ -260,5 +260,14 @@ function ding_libs_libraries_info() {
+      'name' => 'Symfony YAML',
+      'vendor url' => 'https://symfony.com/doc/current/components/yaml.html',
+      'download url' => 'https://github.com/symfony/yaml/archive/v5.2.0.tar.gz',
+      'version' => '5.2',
```
Updated to 3.4 as set by the make file :-)

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 5a12af81ff74d9fddd9e0357e65cf74bfc6b7ee0
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

cableman commented:
The behat test are not related to the changes in this PR.

kasperg commented on modules/ding_config_import/ding_config_import.module:
```diff
@@ -0,0 +1,178 @@
+  $file = $_FILES['files']['tmp_name']['upload'];
+
+  if (libraries_load('yaml')) {
+    $data = file_get_contents($file);
+    try {
+      $values = Yaml::parseFile($data);
```
```suggestion
    try {
      $values = Yaml::parseFile($file);
```

I do not think the current version will work. [`Yaml::parseFile()` takes the path to a file as an argument](https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/Yaml/Parser.php#L55-L82) - not data.

kasperg reviewed (changes requested):
Nice changes. I agree with the prioritization of not implementing some of my suggestions in the current version.

I believe one of the changes will cause the module not to work. Please look into this.

Also consider my suggestion regarding `_array` instead of `array`.

cableman commented:
Updated the PR based on last review changes:

1. Rebased with core
2. Renamed commit id
3. Change `array`  to `_array`
4. Fixed parsing function and testet it (was broken after last review)
